File: Excel/900-999/953/953 Challenge.R
library(tidyverse)
library(readxl)

path <- "900-999/953/953 Streak Identification.xlsx"
input <- read_excel(path, range = "A2:C24")
test <- read_excel(path, range = "E2:G6")

result = input %>%
  mutate(
    group = cumsum(MachineID != lag(MachineID, default = first(MachineID)))
  ) %>%
  group_by(group) %>%
  mutate(
    active = case_when(
      Value >= 80 ~ T,
      Value < 80 & lag(Value) >= 80 & lead(Value) >= 80 ~ T,
      TRUE ~ F
    )
  ) %>%
  ungroup() %>%
  filter(active) %>%
  mutate(
    subgroup = cumsum(TimeID - lag(TimeID, default = first(TimeID)) != 1)
  ) %>%
  mutate(
    StartID = min(TimeID),
    EndID = max(TimeID),
    .by = c(group, subgroup)
  ) %>%
  select(MachineID, StartID, EndID) %>%
  distinct()

all.equal(result, test)
# TRUE

File: Excel/900-999/953/953 Challenge.py
import pandas as pd

path = "900-999/953/953 Streak Identification.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=23, skiprows=1)
test = pd.read_excel(path, usecols="E:G", nrows=4, skiprows=1).rename(columns={"MachineID.1": "MachineID"})

result = input.copy()
result["group"] = (
	result["MachineID"]
	.ne(result["MachineID"].shift(fill_value=result["MachineID"].iloc[0]))
	.cumsum()
)
value_prev = result.groupby("group")["Value"].shift(1)
value_next = result.groupby("group")["Value"].shift(-1)
result["active"] = (
	(result["Value"] >= 80)
	| ((result["Value"] < 80) & (value_prev >= 80) & (value_next >= 80))
)
result = result[result["active"]].copy()
result["subgroup"] = (
	(result["TimeID"] - result["TimeID"].shift(fill_value=result["TimeID"].iloc[0]) != 1)
	.cumsum()
)
result["StartID"] = result.groupby(["group", "subgroup"])["TimeID"].transform("min")
result["EndID"] = result.groupby(["group", "subgroup"])["TimeID"].transform("max")
result = result[["MachineID", "StartID", "EndID"]].drop_duplicates().reset_index(drop=True)

print(result.equals(test))
# Output: True